Reschedule with backoff

greghaynes opened the pull request:
With the alpha scheduling queue we move pods from unschedulable to active on certain events without a backoff. As a result we can cause starvation issues if high priority pods are in the unschedulable queue. Implement a backoff mechanism for pods being moved to active.

```release-note
Take pod backoff in to account when rescheduling a pod for any reason, not just due to a previous scheduling error.
```

Closes #56721

greghaynes committed:
Move scheduling Heap in to scheduler.core.utils

The Heap data structure is useful for our backoff system in addition to
scheduling queue. Move it to somewhere it can be consumed by both
systems and properly export needed names. Also adding unit tests
from client-go/tools/cache/heap.go.

greghaynes committed:
Implement scheduler.util.backoff as a queue

We are going to use PodBackoff for controlling backoff when adding
unschedulable pods back to the active scheduling queue. In order to do
this more easily, limit the interface for PodBackoff to only this struct
(rather than exposing BackoffEntry) and change the backing expiry
implementation to be queue based.

resouer commented:
/ok-to-test

guangxuli commented:
/cc

bsalamat commented:
/assign

bsalamat commented on pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -406,6 +422,71 @@ func (p *PriorityQueue) WaitingPodsForNode(nodeName string) []*v1.Pod {
 
+// Run starts the goroutine to pump from podBackoff to activeQ
+func (p *PriorityQueue) Run() {
+	go wait.Until(p.moveNextBackoffToActive, 0, p.stop)
```
Why is this method better than checking backoff time when a pod needs to be moved from unschedulable queue to active queue and if backoff time is larger than zero, we start a go routine that sleeps (by calling `backoff_util.TryWait`) and then moves the pod to the active queue. Your logic looks more complicated to me.

greghaynes replied:
My thinking was that spawning a goroutine potentially per re-scheduled pod could be a scaling issue. I agree that it would be much simpler to do this though, so I'm happy to be told this isn't an issue.

bsalamat commented on pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -169,12 +182,20 @@ type PriorityQueue struct {
+		clock:          clock,
+		activeQ:        util.NewHeap(cache.MetaNamespaceKeyFunc, util.HigherPriorityPod),
 		unschedulableQ: newUnschedulablePodsMap(),
+		podBackoff:     util.CreateDefaultPodBackoffWithClock(clock),
```
I think we should set the max duration to 15 seconds. 60 seconds is too long and given the exponential growth of backoff time, it does not take that many tries to get to 60 seconds.

bsalamat commented on pkg/scheduler/util/backoff_utils.go:
```diff
@@ -59,35 +50,33 @@ func (b *backoffEntry) unlock() {
 
 type PodBackoff struct {
-	perPodBackoff   map[ktypes.NamespacedName]*backoffEntry
+	expiryQ         *Heap
```
Why are you adding these two queues (expiryQ and backoffQ) here? This is very confusing.

greghaynes replied:
The idea is to treat PodBackoff as a queue for the case of finding the next pod we can schedule so we don't need to perform full pass over all pods we know about in order to determine which are done being 'backed off' (this is backoffQ). Once that work is done its pretty simple to change our GC method to also not require a full pass, so I added expiryQ for this.

bsalamat commented on pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -150,14 +156,21 @@ type UnschedulablePods interface {
+	activeQ *util.Heap
 	// unschedulableQ holds pods that have been tried and determined unschedulable.
 	unschedulableQ *UnschedulablePodsMap
+	// podBackoff holds pods which are awaiting a backoff period before being added to activeQ
```
I think the comment would be clearer if you change "holds" to "tracks". Pods awaiting a backoff period will still be hold in the unschedulableQ. The current comment may make readers think that they are moved from unschedulableQ to podBackoff while waiting.

bsalamat commented on pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -406,6 +422,81 @@ func (p *PriorityQueue) WaitingPodsForNode(nodeName string) []*v1.Pod {
+                glog.Error("Did no find real pod")
+            }
+        }
+		p.activeQ.Add(pod)
```
if the pod is nil, this will be incorrect.

bsalamat replied:
We should also do
p.cond.Broadcast() when a pod is added to the activeQ.

bsalamat reviewed (commented):
I spent more time on this PR and found a couple of bugs. I think the logic is a bit too complex and may be error prone. If you have ideas to simplify the logic, please share them. Otherwise, I will think more about it later.

greghaynes commented on pkg/scheduler/util/backoff_utils.go:
```diff
@@ -97,41 +86,134 @@ func (p *PodBackoff) MaxDuration() time.Duration {
+	return true, b.(*backoffEntry).backoffTime()
+}
+
+// Returns the name of the next pod in the backoffQ which has completed backoff.
```
done

greghaynes commented on pkg/scheduler/util/backoff_utils.go:
```diff
@@ -97,41 +86,134 @@ func (p *PodBackoff) MaxDuration() time.Duration {
+	}
+	return false, ktypes.NamespacedName{}
+}
+
```
done

greghaynes commented:
I agree re: the logic gets a bit complex, I don't have any better ideas though. My hope was to separate enough of the complexity out in to backoff_utils in the first two commits in order to mitigate this a bit and make the final scheduler changes less complex, but I am happy to hear other suggestions.

greghaynes commented:
I pulled over the unittests from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/heap_test.go but I am wondering - should we try and get moved to where we can depend on client-go's heap? Aside from the obvious DRY benefits their impl. allows for waiting which may be a nice way to simplify what were doing here.

bsalamat commented:
@greghaynes Actually, I implemented client-go's heap myself to use it in the scheduling queue, but the reason that I didn't was that we needed to have synchronization (mutex locking) across the two data structures in scheduling queue. If we used client-go's heap, we would incur double locking performance hit.

>  their impl. allows for waiting

I am not sure about this part. What kind of waiting are you referring to? If you mean waiting at `Pop()`, our heap has that feature too.

greghaynes commented:
Ah that makes re: not incurring double locking hit.

Not sure why I didn't notice the pop() waiting in our heap when implementing https://github.com/kubernetes/kubernetes/pull/57057/files#diff-ceba89f8fc17231141f24ea89e6a24d7R434 - I should be able to simplify the logic significantly using that

bsalamat commented:
Thanks, @greghaynes! Please be aware that there is important PR to improve performance of the scheduling queue that I sent today: https://github.com/kubernetes/kubernetes/pull/59671. I don't expect your solution to get affected by it, but you may want to take a look just in case.

jayunit100 commented on pkg/scheduler/core/scheduling_queue_test.go:
```diff
@@ -241,12 +250,20 @@ func TestPriorityQueue_Delete(t *testing.T) {
+		t.Error("Expected all unschedulable items to be in backoff.")
+	}
+	clock.t = clock.t.Add(2 * time.Second)
+	q.moveNextBackoffToActive()
```
why are we doing this twice?

greghaynes replied:
Maybe I should add a comment about this. We are adding two pods and moveNextBackoffToActive only moves one pod to the activeQ at a time.

greghaynes commented:
/retest

greghaynes commented:
/test pull-kubernetes-unit

greghaynes commented:
/test pull-kubernetes-integration

greghaynes commented:
I havent given up on this, I am hitting a failure in k8s.io/kubernetes/test/integration/scheduler TestPDBInPreemption. I've chased this down to in the third test case preemption is occuring on node-2 rather than node-1 due to low-pod2 not being seen as a potential victim when we select our preemption node (it is missing form the nodeToVictims map). This, obviously, makes node-2 a better candidate for preemption.

bsalamat commented:
@greghaynes Thanks for looking into the issue. I also looked at the test and it looks to me that node-2 should be picked as well. So, the test may be wrong. I will look into it, but why this PR changes the behavior of preemption? The test passes fine in the master branch.

greghaynes commented:
@bsalamat I agree this PR should not cause any change for this test, I'm still trying to figure out what the cause is. It seems that with this PR low-pod2 is not in the nodeInfo for node 2 when the scheduler performs preemption for the third test case.

bsalamat commented:
@greghaynes FYI, I sent out #62933 to fix the test.

greghaynes commented:
@bsalamat Thanks, I verified your PR fixes the issue I was hitting. I think the change in this PR which caused the bug to surface is that rescheduling a pod after a failure (which is the codepath for performing preemption) is now always an async process. This made me notice that in the later test cases the preempting pod does end up experiencing backoff which we have to wait for before the test can pass. I'm still working on a fix for this behavior change.

bsalamat commented:
@greghaynes do you plan to work on this PR again? Those failing tests must be passing now.

greghaynes commented:
Great! Yes - I'll try to find some time this weekend/early next week to rebase and cleanup.

bsalamat commented on pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -400,35 +430,22 @@ func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
-		if err := p.activeQ.Add(pod); err != nil {
-			glog.Errorf("Error adding pod %v to the scheduling queue: %v", pod.Name, err)
-		}
+		p.ScheduleWithBackoff(pod)
```
Thanks again for this PR. It looks generally good.
I think we must start the backoff timer when a pod is added to the unschedulable queue, not when we move pods from the unschedulable to the activeQ. This would ensure that a pod that has been waiting for a long time in the unschedulable queue is not subject to further backoff when the cluster has the resources to run it.

bsalamat replied:
I thought more about this. While we still want to start the timer from the moment a pod goes to the unschedulable queue, we must ensure that if a move from the unschedulable queue to the active queue is requested, pods in the unschedulable queue go to the active queue after their backoff time expires.

One idea to address the problem is:
1. Attach a backoff timestamp to every pod that goes to the unschedulable queue.
2. Change ActiveQ to sort pods first by their "seconds to expire" and then by their priority.
3. Move all the pods, including those pods whose backoff time has not expired yet, to the ActiveQ when a move request comes in.

greghaynes replied:
Good point and I like your idea, that should simplify this as well. Thanks!

I wont be able to get to this until early next week but I'll get this fixed up then.

greghaynes replied:
I played around with this a bit and, in addition to what you say about needing to start backoff at the right time, I think more complex sorting/popping off of activeQ is needed if we want priority to effect scheduling order. If we order by backoff expiry in activeQ then this is the order we will end up scheduling if we simply activeQ.pop() to schedule (even if we have a secondary sort of priority). What we want is a priority ordered queue of everything which has completed backoff.

The naive solution for this is fairly computationally expensive - for every schedule attempt do a full activeQ scan to filter based on backoff expiry and sort this result by priority. My thinking right now is to keep a separate heap of pods ordered by backoff and keep activeQ as a pod priority ordered queue. Whenever we try to schedule we first pop all pods which have completed exipiry off our backoff queue and in to activeQ then we pop the head of activeQ (i.e. activeQ becomes a cache of priority ordered pods which have completed backoff).

Curious if you agree or if you think I am missing something?

bsalamat replied:
Sorry for the delay. I just noticed that you have replied.
I agree with you. I think your solution works well. We need to have a separate heap for pods in their backoff period. The heap is sorted by the backoff expiry. ActiveQ remains as is, i.e. sorted by priority. The algorithm is roughly like this:

1. New pods that the scheduler has never tried to schedule go directly to the ActiveQ (this logic already exists).
1. Pods that are attempted and are unschedulable are moved to the unschedulable queue and get a backoff timestamp.
1. When a move request arrives, pods whose backoff period is expired, are moved to the ActiveQ directly. The rest go to the BackoffQ.
1. A separate go routine checks periodically whether the head of the backoffQ is expired. If so, it moves it to the ActiveQ and checks the next head.

greghaynes committed:
Reschedule with backoff

With the alpha scheduling queue we move pods from unschedulable to
active on certain events without a backoff. As a result we can cause
starvation issues if high priority pods are in the unschedulable queue.
Implement a backoff mechanism for pods being moved to active.

Closes #56721

bsalamat commented on pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -292,7 +360,11 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(pod *v1.Pod) error {
+		return fmt.Errorf("pod is already present in the backoffQ")
+	}
 	if !p.receivedMoveRequest && isPodUnschedulable(pod) {
+		p.backoffPod(pod)
```
I think this should be changed to:

```go

if !p.receivedMoveRequest && isPodUnschedulable(pod) {
	p.backoffPod(pod)
	p.unschedulableQ.addOrUpdate(pod)
	p.addNominatedPodIfNeeded(pod)
	return nil
}

// If a move request has been received and the pod is subject to backoff, move it to the BackoffQ.
if p.isPodBackingOff(pod) && isPodUnschedulable(pod) {
	if err := p.podBackoffQ.Add(pod); err != nil {
		glog.Errorf("Error adding pod %v to the backoff queue: %v", pod.Name, err)
	} else {
		p.addNominatedPodIfNeeded(pod)
	}
	return err
}
err := p.activeQ.Add(pod)
```

bsalamat commented on pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -343,8 +448,15 @@ func isPodUpdated(oldPod, newPod *v1.Pod) bool {
 func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
 	p.lock.Lock()
 	defer p.lock.Unlock()
+	_, exists, _ := p.activeQ.Get(newPod)
```
I think we should use `oldPod` here and the line below. Although they should be practically the same given the current "key" function, but we might change the key function in the future.

bsalamat commented on pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -343,8 +448,15 @@ func isPodUpdated(oldPod, newPod *v1.Pod) bool {
+	_, boExists, _ := p.podBackoffQ.Get(newPod)
 	// If the pod is already in the active queue, just update it there.
-	if _, exists, _ := p.activeQ.Get(newPod); exists {
+	if exists || boExists {
```
This logic is a bit risky and hard to read. How about this:

```go
// If the pod is already in the active queue, just update it there.
if _, exists, _ := p.activeQ.Get(newPod); exists {
	p.updateNominatedPod(oldPod, newPod)
	err := p.activeQ.Update(newPod)
	return err
}

// If the pod is in the backoff queue, update it there.
if _, exists, _ := p.podBackoffQ.Get(newPod); exists {
	p.updateNominatedPod(oldPod, newPod)
	err := p.backoffQ.Update(newPod)
	return err
}
```

bsalamat reviewed (commented):
Thanks, @greghaynes! I think the logic is generally correct. I focused this review on the scheduling_queue.go. I will review the rest of the code soon as well.

Please feel free to ping me on slack @bsalamat. I cannot keep up with github notification emails. So, I may not see your new changes quickly.

waynepeking348 commented:
Hi, @greghaynes are you still working on this PR? It seems there still remains a little work before it can merged.

greghaynes commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

looks like network error

greghaynes commented:
/test pull-kubernetes-unit

bsalamat commented on pkg/scheduler/util/backoff_utils.go:
```diff
@@ -37,56 +37,58 @@ func (realClock) Now() time.Time {
 
-func (b *BackoffEntry) wait(maxDuration time.Duration) {
+// backoffAndWait Blocks until this entry has completed backoff
+func (b *backoffEntry) backoffAndWait(maxDuration time.Duration) {
```
Is this used anywhere?

greghaynes replied:
Nope, removed

bsalamat commented on pkg/scheduler/util/backoff_utils.go:
```diff
@@ -111,34 +113,107 @@ func CreatePodBackoff(defaultDuration, maxDuration time.Duration) *PodBackoff {
+
+// GetBackoffTime returns the time that podID completes backoff
+func (p *PodBackoff) GetBackoffTime(podID ktypes.NamespacedName) (time.Time, bool) {
+	rawBe, exists, _ := p.expiryQ.GetByKey(podID.String())
```
we need mutex locking here.

bsalamat commented on pkg/scheduler/util/backoff_utils.go:
```diff
@@ -111,34 +113,107 @@ func CreatePodBackoff(defaultDuration, maxDuration time.Duration) *PodBackoff {
+}
+
+// BackoffPod updates the backoff for a podId and returns the duration until backoff completion
+func (p *PodBackoff) BackoffPod(podID ktypes.NamespacedName) time.Duration {
```
Could you please add unit-tests for these new functions that you have added to `PodBackoff`?

greghaynes replied:
I added tests for ClearPodBackoff and TryBackoffAndWait, I renamed TestBackoff to TestBackoffPod so it it more clear this is a unit test for this case. GetBackoffTime is also pretty well covered in this case.

bsalamat reviewed (commented):
Thanks a lot, @greghaynes!
Most of my comments are minor ones.
I will take a closer look at the backoff_utils.go again.

greghaynes committed:
Check for shutdown in TryBackoffAndWait

greghaynes commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -88,6 +90,10 @@ type FIFO struct {
 var _ = SchedulingQueue(&FIFO{}) // Making sure that FIFO implements SchedulingQueue.
 
+// Run is a noop here but needed to implement SchedulingQueue
+func (f *FIFO) Run() {
```
Nope

greghaynes commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -392,23 +512,37 @@ func isPodUpdated(oldPod, newPod *v1.Pod) bool {
+		}
+
+		// If the pod is in the backoff queue, update it there.
+		if _, exists, _ := p.podBackoffQ.Get(oldPod); exists {
```
Yep, I agree - done.

greghaynes commented:
/test pull-kubernetes-verify

unrelated networking timeout

greghaynes commented:
/test pull-kubernetes-unit

greghaynes commented:
/retest

bsalamat reviewed (approved):
/lgtm
/approve

Thanks a lot, @greghaynes! This is a complex PR. I really appreciate the time and effort you put on this!

greghaynes commented:
@bsalamat No problem, thanks for the reviews!
